Fix upgrade of older saving format

sledgehammer999 opened the pull request:
Even though we had a saving format where the torrent queue position
was written at the end of the fastresume filename, there are fastresume
files out there where their filenames end in other charachers besides
digits. eg <hash>.fastresume.1f2r35
This probably comes from a bug in older versions.

As you can see, I opted to not upgrade if there is already a newer fastresume with the correct name. I assume that it is more current. Is that ok?

glassez commented:
>I assume that it is more current. Is that ok?

Sounds correct for me.

sledgehammer999 commented:
I forgot to mention something:
I observed these fastresumes in my BT_backup folder and I was WTF!!!
I don't know if there really out there this kind of artifacts, but it doesn't hurt to account for them IMO.

glassez commented:
Can't it be about QSaveFile (e.g. its temporary files not deleted for some reason)?

sledgehammer999 commented:
>Can't it be about QSaveFile (e.g. its temporary files not deleted for some reason)?

Wow. Yes. That's probably it. According to [docs](https://doc.qt.io/qt-5/qtemporaryfile.html#QTemporaryFile):
>Constructs a QTemporaryFile using as file template the application name returned by QCoreApplication::applicationName() (otherwise qt_temp) followed by ".XXXXXX". The file is stored in the system's temporary directory, QDir::tempPath().

aka dot followed by 6 characters. And that's exactly what I had in my BT_backup folder. Maybe those files are artifacts after a crash during fastresume saving.

If that's the case, we should continue accounting for it even after we remove the upgrade code.
I probably need to update the code comment about this finding.

sledgehammer999 commented:
@glassez I updated this. I added a bit of different approach if the fastresume is an artifact of QSaveFile. I assume that these are always corrupt/incomplete fastresumes and always remove them.

glassez commented on src/app/upgrade.h:
```diff
@@ -112,14 +112,28 @@ bool upgradeResumeFile(const QString &filepath, const QVariantHash &oldTorrent =
+        // be a corrupted/incomplete fastresume.
+        // NOTE: When the upgrade code is removed, we must continue to perform cleanup of
+        // QSaveFile/QTemporaryFile artifacts
         queuePosition = rx.cap(2).toInt();
```
I would just check rx.cap(2) with regex instead of misunderstandable expression below.

sledgehammer999 replied:
The expression below is explained in the comment above. I like it better this way.

sledgehammer999 replied:
No, because:
1. Assume this name `hash.fastresume.000001`
2. `queuePosition` will be `1`. But old v3_3 format would have never written it like that. It would have written it as `hash.fastresume.1`
3. In case it isn't clear I test for 6 chars because a name like `hash.fastresume.13bafr` would result in `queuePosition == 13`

glassez replied:
>Assume this name hash.fastresume.000001

So what's wrong with my expression? It matches this filename as artifact and the file will be deleted.

Chocobo1 replied:
I think it can be simplified to:
```c++
queuePosition = rx.cap(2).toInt();  // returns 0 when conversion failed
if (!((99999 < queuePosition) && (queuePosition <= 999999))) {
    Utils::Fs::forceRemove(filepath);
    return true;
}
```

sledgehammer999 replied:
Damn it. You're right. @glassez I read your number as six nines not as five nines.
@Chocobo1 I think your condition will also delete valid v3_3 files. eg `hash.fastresume.12345`

I am pushing a version with @glassez's suggestion.

glassez commented on src/app/upgrade.h:
```diff
@@ -112,14 +112,28 @@ bool upgradeResumeFile(const QString &filepath, const QVariantHash &oldTorrent =
+        if ((rx.cap(2).size() == 6)
+            && !((queuePosition > 99999) && (queuePosition <= 999999)))
+            artifact = true;
         v3_3 = true;
```
It should be in "else" branch since it's separate case.

sledgehammer999 commented:
1. New PR title
2. Split PR into 2 commits. I think they do separate enough things to award separate commits.
3. New long commit message
4. New code comment (almost the same as commit message)
5. Simpler approach in deleting non-commited fastresumes
